core.ts

```diff
@@ -0,0 +1,1039 @@
+	const results = []
+	for (const key in obj) {
+		if (obj.hasOwnProperty(key)) {
+			results.push(fn(obj[key], key))
+		}
+	}
+	return results
+}
+
+const requiredUtilMethods = ['isString', 'isNumber', 'isBoolean', 'isObject', 'isArray', 'isDate']
+const isUtil = (obj: any): obj is PinejsClientCoreFactory.Util => {
+	if (obj == null) {
+		return false
+	}
+	for (const method of requiredUtilMethods) {
+		if (obj[method] == null) {
+			return false
+		}
+	}
+	return true
+}
+const isPromiseRejector = (obj: any): obj is PinejsClientCoreFactory.PromiseRejector => {
+	return obj != null && obj.reject != null
+}
+const isValidOption = (key: string): key is keyof PinejsClientCoreFactory.ODataOptions => {
+	return key === '$filter' ||
+		key === '$expand' ||
+		key === '$orderby' ||
+		key === '$top' ||
+		key === '$skip' ||
+		key === '$select'
+}
+
+// Workaround the fact that `setInterval` returns a different type in nodejs vs browsers
+// TODO: typescript 2.8 will introduce `ReturnType` as a better way to do this
+const _setIntervalResult = false as true && setInterval(() => {}, 0)
+
+type PollOnObj = {
+	unsubscribe: () => void
+}
+class Poll<PromiseResult extends PromiseLike<number | PinejsClientCoreFactory.AnyObject | PinejsClientCoreFactory.AnyObject[]> = Promise<number | PinejsClientCoreFactory.AnyObject | PinejsClientCoreFactory.AnyObject[]>> {
+	private subscribers: {
+		error: Array<(response: PromiseResult) => void>,
+		data: Array<(err: any) => void>,
+	} = {
+		error: [],
+		data: [],
+	}
+
+	private stopped = false
+	private pollInterval?: typeof _setIntervalResult
+
+	private requestFn: null | (() => PromiseResult)
+
+	constructor(
+		requestFn: () => PromiseResult,
+		private intervalTime = 10000
+	) {
+		this.requestFn = requestFn
+		this.start()
+	}
+
+	setPollInterval(intervalTime: number) {
+		this.intervalTime = intervalTime
+		this.stop()
+		this.start()
+	}
+
+	runRequest() {
+		if (this.stopped || this.requestFn == null) {
+			return
+		}
+		this.requestFn()
+		.then((response) => {
+			if (this.stopped) {
+				return
+			}
+
+			// Catch errors in event subscribers so that they don't trigger
+			// the 'catch' below, and that subsequent subscribers will still
+			// be called
+			this.subscribers.data.forEach((fn) => {
+				try {
+					fn(response)
+				} catch (error) {
+					console.error('pinejs-client error: Caught error in data event subscription:', error)
+				}
+			})
+
+			return null
+		}, (err: any) => {
+			if (this.stopped) {
+				return
+			}
+
+			this.subscribers.error.forEach((fn) => {
+				try {
+					fn(err)
+				} catch (error) {
+					console.error('pinejs-client error: Caught error in error event subscription:', error)
+				}
+			})
+
+			return null
+		})
+	}
+
+	on(name: 'error', fn: (response: PromiseResult) => void): PollOnObj
+	on(name: 'data', fn: (err: any) => void): PollOnObj
+	on(name: keyof Poll['subscribers'], fn: (value: any) => void): PollOnObj {
```

Comment:
The TS type system is insanely descriptive.

Though I think you should be able to simply set this to `name: 'error' | 'data'`? Any reason not to? Is that intended to be a catch-all for all events defined in "subscribers"?

Reply:
As it stands `keyof Poll['subscribers']` is the same as `'error' | 'data'` but doing it this way is accurate to the true type and guarantees that it will stay in sync.  It also means that if we were to remove the `error` or `data` keys from `subscribers` then the relevant overload would start to throw a type error as either `'error'` or `'data` would no longer be a subset.  Overall it just ends up with greater type safety

Also the reason we need this more generic form for the actual function is because overloads must be compatible with the main function definition